typescript/src/lib/bitcoin/address.ts

```diff
@@ -0,0 +1,121 @@
+import { address as btcjsaddress, payments } from "bitcoinjs-lib"
+import { Hex } from "../utils"
+import { BitcoinNetwork, toBitcoinJsLibNetwork } from "./network"
+
+/**
+ * Creates the Bitcoin address from the public key. Supports SegWit (P2WPKH) and
+ * Legacy (P2PKH) formats.
+ * @param publicKey - Public key used to derive the Bitcoin address.
+ * @param bitcoinNetwork - Target Bitcoin network.
+ * @param witness - Flag to determine address format: true for SegWit (P2WPKH)
+ *        and false for Legacy (P2PKH). Default is true.
+ * @returns The derived Bitcoin address.
+ */
+export function publicKeyToAddress(
+  publicKey: Hex,
+  bitcoinNetwork: BitcoinNetwork,
+  witness: boolean = true
+): string {
+  const network = toBitcoinJsLibNetwork(bitcoinNetwork)
+
+  if (witness) {
+    // P2WPKH (SegWit)
+    return payments.p2wpkh({ pubkey: publicKey.toBuffer(), network }).address!
+  } else {
+    // P2PKH (Legacy)
+    return payments.p2pkh({ pubkey: publicKey.toBuffer(), network }).address!
+  }
+}
+
+/**
+ * Converts a public key hash into a P2PKH/P2WPKH address.
+ * @param publicKeyHash Public key hash that will be encoded. Must be an
+ *        unprefixed hex string (without 0x prefix).
+ * @param witness If true, a witness public key hash will be encoded and
+ *        P2WPKH address will be returned. Returns P2PKH address otherwise
+ * @param bitcoinNetwork Network the address should be encoded for.
+ * @returns P2PKH or P2WPKH address encoded from the given public key hash
+ * @throws Throws an error if network is not supported.
+ */
+function publicKeyHashToAddress(
+  publicKeyHash: string,
+  witness: boolean,
+  bitcoinNetwork: BitcoinNetwork
+): string {
+  const hash = Buffer.from(publicKeyHash, "hex")
+  const network = toBitcoinJsLibNetwork(bitcoinNetwork)
+  return witness
+    ? payments.p2wpkh({ hash, network }).address!
+    : payments.p2pkh({ hash, network }).address!
+}
+
+/**
+ * Converts a P2PKH or P2WPKH address into a public key hash. Throws if the
+ * provided address is not PKH-based.
+ * @param address P2PKH or P2WPKH address that will be decoded.
+ * @param bitcoinNetwork Network the address should be decoded for.
+ * @returns Public key hash decoded from the address. This will be an unprefixed
+ *        hex string (without 0x prefix).
+ */
+function addressToPublicKeyHash(
+  address: string,
+  bitcoinNetwork: BitcoinNetwork
+): string {
+  const network = toBitcoinJsLibNetwork(bitcoinNetwork)
+
+  try {
+    // Try extracting hash from P2PKH address.
+    const hash = payments.p2pkh({ address: address, network }).hash!
+    return hash.toString("hex")
```

Comment:
We could use `Hex` util for conversion.
```suggestion
    return Hex.from(hash).toString()
```

I won't mention it again for other places we deal with hex :pray: